File: RIONet/socket_handlers/SenderSocket.java
package RIONet.socket_handlers;

import java.net.Socket;
import java.io.IOException;
import java.io.DataOutputStream;

import RIONet.Packet;

public class SenderSocket {
    private Socket sock;
    private DataOutputStream outStream;

    public SenderSocket() {
    }

    /**
     * connects the sender to a listener on the given ip and port
     *
     * @param ip   the ip of the listener to connect to
     * @param port the port of the listener to connect to
     */
    public void connect(String ip, int port) throws IOException {
        sock = new Socket(ip, port);
        outStream = new DataOutputStream(sock.getOutputStream());
    }

    /**
     * sends data to the listener, the data must be wrapped around a DataObject
     *
     * @param data DataObject the data to send
     */
    public void sendData(Packet packet) throws IOException, SocketHandlerException {
        if (outStream == null)
            throw new SocketHandlerException("Must first astablish a connection to listener before sending!");
        outStream.write(packet.serialize());
    }
}

File: PUNet/socket_handlers/SenderSocket.py
import socket

from PUNet.Packet import Packet
from .SockethandlerException import SockethandlerException


class SenderSocket:
    """A socket handler for sending data.
    Implements a client that sends data to listeners.
    """
    def __init__(self) -> None:
        """Creates a new sender socket
        """
        self.connected = False
        self.sock: socket.socket = socket.socket(
            socket.AF_INET, socket.SOCK_STREAM
        )

    def connect(self, ip: str, port: int) -> None:
        """Connects the sender to a listener on a specified ip and port.

        :param ip: the ip of the listener
        :type ip: str
        :param port: the port of the listener
        :type port: int
        """
        self.sock.connect((ip, port))
        self.connected = True

    def send_data(self, packet: Packet) -> None:
        """Sends a packet to a the listener.

        :param data: the packet
        :type data: Packet
        :raises SockethandlerException: if the sender isn't connected to any listener
        """
        if self.connected:
            self.sock.send(packet.serialize())
        else:
            raise SockethandlerException(
                "Must first astablish a connection to listener before sending!"
            )

    def is_connected(self) -> bool:
        return self.connected
